leases/services/invoice.py

```diff
@@ -0,0 +1,242 @@
+        self.failure_count += 1
+
+    def fail_order(self, order: Order, message: str,) -> None:
+        """Set an order to ERROR status and append it to the failed_order list"""
+        order.comment = f"{today().date()}: {message}"
+        order.save(update_fields=["comment"])
+        order.set_status(OrderStatus.ERROR, f"Lease renewing failed: {message}")
+        self.failed_orders.append({order.id: message})
+        self.failure_count += 1
+
+    def send_email(self, order, email):
+        # If the profile doesn't have an associated email
+        # or if it has an example.com address, set it as failed
+        if email and "example.com" not in email:
+            try:
+                approve_order(order, email, self.due_date, self.request)
+                self.successful_orders.append(order.id)
+            except (
+                AnymailError,
+                OSError,
+                ValidationError,
+                VenepaikkaGraphQLError,
+            ) as e:
+                self.fail_order(order, str(e))
+        else:
+            self.fail_order(order, _("Missing customer email"))
+
+    def send_invoices(self) -> Dict[str, List]:  # noqa: C901
+        self.successful_orders = []
+        self.failed_orders = []
+        self.failed_leases = []
+
+        self.failure_count = 0
+
+        leases = self.get_valid_leases(self.season_start)
+
+        # Fetch all the profiles from the Profile service
+        profiles = ProfileService(self.request).get_all_profiles()
+
+        for lease in leases:
+            order = None
+            if self.failure_count >= self.MAXIMUM_FAILURES:
+                raise AutomaticInvoicingError(
+                    _(
+                        f"Limit of failures reached: {self.failure_count} elements failed"
+                    )
+                )
+
+            try:
+                with transaction.atomic():
+                    new_lease = self.create_new_lease(
+                        lease, self.season_start, self.season_end
+                    )
+                    customer = new_lease.customer
+
+                    # Get the associated product
+                    product = self.get_product(new_lease)
+
+                    # If no product is found, the billing can't be done
+                    if not product:
+                        self.fail_lease(new_lease, _("No suitable product found"))
+                        continue
+
+                    try:
+                        order = Order.objects.create(
+                            customer=customer, lease=new_lease, product=product
+                        )
+
+                        self.send_email(
+                            order, profiles.get(customer.id).email,
+                        )
+                    except ValidationError as e:
+                        self.fail_lease(new_lease, str(e))
+            except DatabaseError as e:
+                lease.refresh_from_db()
```

Comment:
What is the expected case where DatabaseError would be thrown? Would it make sense to catch a more specific type of DatabaseException, like IntegrityError? DatabaseError could be something that is not recoverable.

Reply:
Not really sure. I was just trying to catch all the possible errors, but you make a good point that it could be not recoverable. Let me take a look at that.

Reply:
Well if it is a serious database error (connection lost or whatever), then it would most likely throw another exception in refresh_from_db, so I think the code is ok as-is, maybe log error stacktrace?